### DISCUSSION TITLE
golang driver query API

### DISCUSSION THREAD

**Discussion by fmoor** [geldata/gel#1917] [General]
I would like feedback on the golang driver query API.

Currently the query API is modeled after the python and java script implementations with adjustments for [serialization](https://github.com/edgedb/edgedb/discussions/1915) and [timeouts](https://blog.golang.org/context). `args` may be a single `map[string]interface{}` when the query uses named arguments or any number of scalar values when the query uses numbered arguments.

```go
type Client struct{}
func (c *Client) Execute(ctx context.Context, query string) error
func (c *Client) Query(ctx context.Context, query string, out interface{}, args ...interface{}) error
func (c *Client) QueryOne(ctx context.Context, query string, out interface{}, args ...interface{}) error
func (c *Client) QueryJSON(ctx context.Context, query string, args ...interface{}) ([]byte, error)
func (c *Client) QueryOneJSON(ctx context.Context, query string, args ...interface{}) ([]byte, error)
```

# Examples
Arguments:
```go
type Movie struct {
    Title string `edgedb:"title"`
    Year  int64  `edgedb:"year"`
}

query := `
    SELECT Movie { title, year }
    FILTER .year = <int64>$0
`

var results []Movie
client.Query(ctx, query, &results, 2017)

fmt.Println(result)
```
Nested objects and collections:
```go
type Person struct {
    FirstName string `edgedb:"first_name"`
    LastName  string `edgedb:"last_name"`
}

type Movie struct {
    Title    string   `edgedb:"title"`
    Year     int64    `edgedb:"year"`
    Director Person   `edgedb:"director"`
    Actors   []Person `edgedb:"actors"`
}

query := `
SELECT Movie {
        title,
        year,
        director: {
                first_name,
                last_name
        },
        actors: {
                first_name,
                last_name
        }
}
`

var results []Movie
client.Query(ctx, query, &results)

fmt.Println(results)
```

# Timeouts & context.Context
The [go blog](https://blog.golang.org/context) indicates that `context.Context` should be passed down the call stack so that request specific timeouts can be honored. Some database drivers adopt this construct ([mongo](https://pkg.go.dev/github.com/mongodb/mongo-go-driver@v1.2.1/mongo)) others do not ([redigo](https://pkg.go.dev/github.com/gomodule/redigo@v1.8.2/redis), [mgo](https://pkg.go.dev/github.com/globalsign/mgo)) and some do both ([pgx](https://pkg.go.dev/github.com/jackc/pgx), [pg](https://pkg.go.dev/github.com/go-pg/pg)).

**Comment by 1st1** [comment]
> Example

Can you show how a more elaborate query would work?  E.g. how would you unpack nested objects and collections of nested objects into nested structs?

**Reply by fmoor** [reply]
I added another example. Let me know if its still missing something.

**Reply by 1st1** [reply]
This is pretty cool. Does this already work?

**Reply by fmoor** [reply]
Yes, everything listed here already works.
